Xiao/DisarmSecurityArtifact

ciaozhang opened the pull request:
Remove the signatures from JWS and authentication tag from JWE when PII is on.
1. For SecurityToken: A new interface is introduced: Microsoft.IdentityModel.Logging. ISafeLogSecurityArtifact with a single method UnsafeToString(). SecurityToken will implementations (JsonWebToken, SAML, etc.) the interface and will provide an implementation of UnsafeToString().
2. For type of object that is not able modified (such as ‘string’): Struct SecurityArtifact can be used to wrap the object and provide a callback to Log appropriately.

ciaozhang committed:
Disarming Authentication Artifacts

TimHannMSFT commented on src/Microsoft.IdentityModel.JsonWebTokens/JsonWebToken.cs:
```diff
@@ -579,14 +579,26 @@ private void ReadToken(string encodedJson)
+            if (!IsEncrypted)
+                return Header.ToString() + "." + Payload.ToString();
+            else
+                return Header.ToString()  + "." + EncryptedKey + "." + InitializationVector + "." + Ciphertext;
```
####                 return Header.ToString()  + "." + EncryptedKey + "." + InitializationVector + "." + Ciphertext;

---

Have you evaluated StringBuilder in this case? Usually when you're concatonating more than ~3 strings StringBuilder wins. #Resolved

TimHannMSFT replied:
For three string parts:
> |                          Method |       Mean |     Error |    StdDev |     Median |   Gen0 | Allocated |
|-------------------------------- |-----------:|----------:|----------:|-----------:|-------:|----------:|
| ConcatStringsUsingStringBuilder | 27.4693 ns | 1.1695 ns | 3.2793 ns | 26.5547 ns | 0.0242 |     152 B |
|          ConcatStringsUsingPlus |  0.0000 ns | 0.0000 ns | 0.0000 ns |  0.0000 ns |      - |         - |

Interesting, seems like just plussing them together still wins at 7....Benchmark.net isn't able to distinguish a difference between that and an empty method.
> |                           Method |       Mean |     Error |    StdDev |   Gen0 | Allocated |
|--------------------------------- |-----------:|----------:|----------:|-------:|----------:|
| ConcatStringsUsingStringBuilder7 | 61.3046 ns | 1.2456 ns | 1.4828 ns | 0.0459 |     288 B |
|          ConcatStringsUsingPlus7 |  0.0000 ns | 0.0000 ns | 0.0000 ns |      - |         - |

I forced in a ToString() to make sure it was actually sampling but got a similar result so seems like your code is good here.
> |                           Method |     Mean |    Error |    StdDev |   Median |   Gen0 | Allocated |
|--------------------------------- |---------:|---------:|----------:|---------:|-------:|----------:|
| ConcatStringsUsingStringBuilder7 | 97.53 ns | 3.743 ns | 10.370 ns | 93.10 ns | 0.0790 |     496 B |
|          ConcatStringsUsingPlus7 | 12.99 ns | 0.601 ns |  1.752 ns | 12.45 ns | 0.0217 |     136 B |

victorm-hernandez replied:
@TimHannMSFT , I am curious, what did you use to do your benchmark? I would like to use that tool.

About the string concatenation at hand, I found that in these type of scenarios where the string concatenation happens in a single line, the compiler adds the required optimizations such that there are no immutable strings waste here.

In this case the optimization looks like
``` csharp
        string[] array = new string[7];
        array[0] = header.ToString();
        array[1] = ".";
        array[2] = EncryptedKey;
        array[3] = ".";
        array[4] = InitializationVector;
        array[5] = ".";
        array[6] = Ciphertext;
        return string.Concat(array);
```

TimHannMSFT replied:
https://github.com/dotnet/BenchmarkDotNet

victorm-hernandez replied:
Thanks!

dannybtsai replied:
@TimHannMSFT/@victorm-hernandez  StringBuilder should be used in for loops or dynamics cases. I believe the compiler generates code similar to the string.Concat method where it calculates the total length of all strings first, allocate a memory once then copy all values. It is efficient because it does not need to re-allocate the buffer.

ciaozhang committed:
disarm actor token

TimHannMSFT commented:
@ciaozhang can you add more details to the description as well as a section which can be copied directly into the release notes? I'm happy to help construct if you need help. @jmprieur would also likely be able to help.

---
In reply to: [1524185285](https://github.com/AzureAD/azure-activedirectory-identitymodel-extensions-for-dotnet/pull/2064#issuecomment-1524185285) [](http://example.com/codeflow?ancestors=1524185285)

---
In reply to: [1524185285](https://github.com/AzureAD/azure-activedirectory-identitymodel-extensions-for-dotnet/pull/2064#issuecomment-1524185285) [](http://example.com/codeflow?ancestors=1524185285)

---
In reply to: [1524185285](https://github.com/AzureAD/azure-activedirectory-identitymodel-extensions-for-dotnet/pull/2064#issuecomment-1524185285) [](http://example.com/codeflow?ancestors=1524185285)

brentschmaltz commented on src/Microsoft.IdentityModel.JsonWebTokens/JsonWebToken.cs:
```diff
@@ -579,14 +579,26 @@ private void ReadToken(string encodedJson)
+        public override string ToString()
+        {
+            if (!IsEncrypted)
+                return Header.ToString() + "." + Payload.ToString();
```
####                 return Header.ToString() + "." + Payload.ToString();

---

In most cases EncodedHeader and EncodedPayload is what you want here.
If you fix this bug: https://github.com/AzureAD/azure-activedirectory-identitymodel-extensions-for-dotnet/issues/2065 then you will always get proper value. #Resolved

TimHannMSFT commented on src/Microsoft.IdentityModel.JsonWebTokens/JsonWebTokenHandler.cs:
```diff
@@ -1064,7 +1066,7 @@ internal IEnumerable<SecurityKey> GetContentEncryptionKeys(JsonWebToken jwtToken
                 return unwrappedKeys;
             else
-                throw LogHelper.LogExceptionMessage(new SecurityTokenKeyWrapException(LogHelper.FormatInvariant(TokenLogMessages.IDX10618, keysAttempted, exceptionStrings, jwtToken)));
+                throw LogHelper.LogExceptionMessage(new SecurityTokenKeyWrapException(LogHelper.FormatInvariant(TokenLogMessages.IDX10618, keysAttempted, exceptionStrings, jwtToken.ToString())));
```
####                 throw LogHelper.LogExceptionMessage(new SecurityTokenKeyWrapException(LogHelper.FormatInvariant(TokenLogMessages.IDX10618, keysAttempted, exceptionStrings, jwtToken.ToString())));

---

Does this change do anything? I believe ToString is called by default anyways in string.Format which this eventually calls. #Resolved

sruke replied:
That's correct. LogHelper.FormatInvariant internally invokes string.Format where ToString is called by default

victorm-hernandez replied:
Given that this has no change on the output, and that changing the type of the parameter from JsonWebToken to string will prevent the LogHelper class to be able to handle this type in specific like it does for NonPII or Exception, I would not change it.

victorm-hernandez replied:
Following the same train of thought, this change could be implemented entirely on the LogHelper.FormatInvariant by checking for the type of the object to be SecurityToken and if it is sanitize the output in place.

TimHannMSFT commented on src/Microsoft.IdentityModel.JsonWebTokens/JsonWebTokenHandler.cs:
```diff
@@ -1240,11 +1242,11 @@ private static TokenValidationResult ReadToken(string token, TokenValidationPara
                 var securityToken = validationParameters.TokenReader(token, validationParameters);
                 if (securityToken == null)
-                    throw LogHelper.LogExceptionMessage(new SecurityTokenInvalidSignatureException(LogHelper.FormatInvariant(TokenLogMessages.IDX10510, token)));
+                    throw LogHelper.LogExceptionMessage(new SecurityTokenInvalidSignatureException(LogHelper.FormatInvariant(TokenLogMessages.IDX10510, JwtTokenUtilities.PrepareSecurityArtifact(token))));
```
####                     throw LogHelper.LogExceptionMessage(new SecurityTokenInvalidSignatureException(LogHelper.FormatInvariant(TokenLogMessages.IDX10510, JwtTokenUtilities.PrepareSecurityArtifact(token))));

---

I would prefer we followed the pattern used for PII. Something like, `LogHelper.MarkSecurityArtifact`. This has a couple advantages in my mind
1. It reuses a pattern which is already established in code so it makes it easier to understand what's going on.
2. It allows log level to be taken into account before performing any string operations. We can win from a performance standpoint if we're only doing these string operations (or regEx searches) when we actually go to log. If there's a log informational with a token. Code as written would still do the string manipulation even if the caller had only enabled error logs. #Resolved

TimHannMSFT commented on src/Microsoft.IdentityModel.Logging/SecurityArtifact.cs:
```diff
@@ -0,0 +1,35 @@
+    /// An internal structure that is used to mark an argument as NonPII.
+    /// Arguments wrapped with a NonPII structure will be considered as NonPII in the message logging process.
+    /// </summary>
+    internal struct SecurityArtifact
```
####     internal struct SecurityArtifact

---

Not used, but should be! #Resolved

TimHannMSFT commented on src/Microsoft.IdentityModel.Logging/SecurityArtifact.cs:
```diff
@@ -0,0 +1,35 @@
+    internal struct SecurityArtifact
+    {
+        /// <summary>
+        /// Argument wrapped with a <see cref="NonPII"/> structure is considered as NonPII in the message logging process.
```
####         /// Argument wrapped with a <see cref="NonPII"/> structure is considered as NonPII in the message logging process.

---

Copy/paste artifacts left over, please clean up #Resolved

TimHannMSFT commented on src/Microsoft.IdentityModel.Logging/SecurityArtifact.cs:
```diff
@@ -0,0 +1,35 @@
+        /// <returns><c>Null</c> if the <see cref="Argument"/> is <c>null</c>, otherwise calls <see cref="System.ValueType.ToString()"/> method of the <see cref="Argument"/>.</returns>
+        public override string ToString()
+        {
+            return Argument?.ToString() ?? "Null";
```
####             return Argument?.ToString() ?? "Null";

---

nit: We mix "Null" and "null" throughout the Wilson code base. We should standarize to just one or the other. I'd prefer "null" since that's the casing of in .NET. #Resolved

dannybtsai replied:
There is a Null constant defined in the Utility class;

public const string Null = "null";

We should use it.

ciaozhang replied:
Changed to  "null"

TimHannMSFT commented on src/Microsoft.IdentityModel.Protocols.OpenIdConnect/OpenIdConnectProtocolValidator.cs:
```diff
@@ -346,7 +346,7 @@ protected virtual void ValidateIdToken(OpenIdConnectProtocolValidationContext va
                 catch (Exception ex)
                 {
-                    throw LogHelper.LogExceptionMessage(new OpenIdConnectProtocolException(LogHelper.FormatInvariant(LogMessages.IDX21313, validationContext.ValidatedIdToken), ex));
+                    throw LogHelper.LogExceptionMessage(new OpenIdConnectProtocolException(LogHelper.FormatInvariant(LogMessages.IDX21313, validationContext.ValidatedIdToken.ToString()), ex));
```
Again, it's better to wait until the last minute to do string operations for performance reasons. If our logging helper methods aren't able to handle this we should update the helper/utility methods not each instance where we call.

brentschmaltz commented on src/System.IdentityModel.Tokens.Jwt/JwtSecurityTokenHandler.cs:
```diff
@@ -1141,7 +1141,7 @@ private ClaimsPrincipal ValidateTokenPayload(JwtSecurityToken jwtToken, TokenVal
                 identity.BootstrapContext = jwtToken.RawData;
 
-            LogHelper.LogInformation(TokenLogMessages.IDX10241, jwtToken.RawData);
+            LogHelper.LogInformation(TokenLogMessages.IDX10241, JwtTokenUtilities.PrepareSecurityArtifact(jwtToken.RawData));
```
####             LogHelper.LogInformation(TokenLogMessages.IDX10241, JwtTokenUtilities.PrepareSecurityArtifact(jwtToken.RawData));

---

Could this be SecurityToken.ToString()? #Resolved

TimHannMSFT replied:
IMO neither of those is the ideal way. We want to delay string operations as log as possible. Especially for informational logs which may not ever actually see the light of day. Those would just be wasted CPU cycles.

brentschmaltz commented on src/System.IdentityModel.Tokens.Jwt/JwtSecurityTokenHandler.cs:
```diff
@@ -1151,7 +1151,7 @@ private ClaimsPrincipal CreateClaimsPrincipalFromToken(JwtSecurityToken jwtToken
                 identity.BootstrapContext = jwtToken.RawData;
 
-            LogHelper.LogInformation(TokenLogMessages.IDX10241, jwtToken.RawData);
+            LogHelper.LogInformation(TokenLogMessages.IDX10241, JwtTokenUtilities.PrepareSecurityArtifact(jwtToken.RawData));
```
####             LogHelper.LogInformation(TokenLogMessages.IDX10241, JwtTokenUtilities.PrepareSecurityArtifact(jwtToken.RawData));

---

Could this be SecurityToken.ToString()? #Resolved

TimHannMSFT commented on src/Microsoft.IdentityModel.Protocols.OpenIdConnect/OpenIdConnectProtocolValidator.cs:
```diff
@@ -356,33 +356,33 @@ protected virtual void ValidateIdToken(OpenIdConnectProtocolValidationContext va
+                    throw LogHelper.LogExceptionMessage(new OpenIdConnectProtocolException(LogHelper.FormatInvariant(LogMessages.IDX21314, LogHelper.MarkAsNonPII(JwtRegisteredClaimNames.Aud.ToLowerInvariant()), idToken.ToString())));
 
                 if (!idToken.Payload.Exp.HasValue)
-                    throw LogHelper.LogExceptionMessage(new OpenIdConnectProtocolException(LogHelper.FormatInvariant(LogMessages.IDX21314, LogHelper.MarkAsNonPII(JwtRegisteredClaimNames.Exp.ToLowerInvariant()), idToken)));
+                    throw LogHelper.LogExceptionMessage(new OpenIdConnectProtocolException(LogHelper.FormatInvariant(LogMessages.IDX21314, LogHelper.MarkAsNonPII(JwtRegisteredClaimNames.Exp.ToLowerInvariant()), idToken.ToString())));
 
                 if (!idToken.Payload.Iat.HasValue)
-                    throw LogHelper.LogExceptionMessage(new OpenIdConnectProtocolException(LogHelper.FormatInvariant(LogMessages.IDX21314, LogHelper.MarkAsNonPII(JwtRegisteredClaimNames.Iat.ToLowerInvariant()), idToken)));
+                    throw LogHelper.LogExceptionMessage(new OpenIdConnectProtocolException(LogHelper.FormatInvariant(LogMessages.IDX21314, LogHelper.MarkAsNonPII(JwtRegisteredClaimNames.Iat.ToLowerInvariant()), idToken.ToString())));
 
                 if (idToken.Payload.Iss == null)
-                    throw LogHelper.LogExceptionMessage(new OpenIdConnectProtocolException(LogHelper.FormatInvariant(LogMessages.IDX21314, LogHelper.MarkAsNonPII(JwtRegisteredClaimNames.Iss.ToLowerInvariant()), idToken)));
+                    throw LogHelper.LogExceptionMessage(new OpenIdConnectProtocolException(LogHelper.FormatInvariant(LogMessages.IDX21314, LogHelper.MarkAsNonPII(JwtRegisteredClaimNames.Iss.ToLowerInvariant()), idToken.ToString())));
```
Same comment here about agressive ToString() calls. Best to avoid. I'm curious if this has any functional difference at runtime.

brentschmaltz commented on src/System.IdentityModel.Tokens.Jwt/JwtSecurityTokenHandler.cs:
```diff
@@ -1291,7 +1291,7 @@ private JwtSecurityToken ValidateSignature(string token, JwtSecurityToken jwtTok
             {
                 if (validationParameters.RequireSignedTokens)
-                    throw LogHelper.LogExceptionMessage(new SecurityTokenInvalidSignatureException(LogHelper.FormatInvariant(TokenLogMessages.IDX10504, token)));
+                    throw LogHelper.LogExceptionMessage(new SecurityTokenInvalidSignatureException(LogHelper.FormatInvariant(TokenLogMessages.IDX10504, JwtTokenUtilities.PrepareSecurityArtifact(token))));
```
####                     throw LogHelper.LogExceptionMessage(new SecurityTokenInvalidSignatureException(LogHelper.FormatInvariant(TokenLogMessages.IDX10504, JwtTokenUtilities.PrepareSecurityArtifact(token))));

---

Could you use the JwtSecurityToken passed in the parameter? #Resolved

TimHannMSFT replied:
Would be nice if the method header described how these were different. Coming to this cold, it's not clear why both would exist.

brentschmaltz commented on src/System.IdentityModel.Tokens.Jwt/JwtSecurityTokenHandler.cs:
```diff
@@ -1347,7 +1347,7 @@ private JwtSecurityToken ValidateSignature(string token, JwtSecurityToken jwtTok
                         if (ValidateSignature(encodedBytes, signatureBytes, key, jwtToken.Header.Alg, jwtToken, validationParameters))
                         {
-                            LogHelper.LogInformation(TokenLogMessages.IDX10242, token);
+                            LogHelper.LogInformation(TokenLogMessages.IDX10242, JwtTokenUtilities.PrepareSecurityArtifact(token));
```
####                             LogHelper.LogInformation(TokenLogMessages.IDX10242, JwtTokenUtilities.PrepareSecurityArtifact(token));

---

Could you use the JwtSecurityToken passed in the parameter? #Resolved

brentschmaltz commented on src/Microsoft.IdentityModel.JsonWebTokens/JsonWebTokenHandler.cs:
```diff
@@ -1573,7 +1575,7 @@ private static JsonWebToken ValidateSignature(JsonWebToken jwtToken, TokenValida
 #endif
                         {
-                            LogHelper.LogInformation(TokenLogMessages.IDX10242, jwtToken.EncodedToken);
+                            LogHelper.LogInformation(TokenLogMessages.IDX10242, JwtTokenUtilities.PrepareSecurityArtifact(jwtToken.EncodedToken));
```
####                             LogHelper.LogInformation(TokenLogMessages.IDX10242, JwtTokenUtilities.PrepareSecurityArtifact(jwtToken.EncodedToken));

---

Could you use jwtToken.ToString() #Resolved

TimHannMSFT commented on src/Microsoft.IdentityModel.Protocols.OpenIdConnect/OpenIdConnectProtocolValidator.cs:
```diff
@@ -356,33 +356,33 @@ protected virtual void ValidateIdToken(OpenIdConnectProtocolValidationContext va
                 // sub is required in OpenID spec; but we don't want to block valid idTokens provided by some identity providers
                 if (RequireSub && (string.IsNullOrWhiteSpace(idToken.Payload.Sub)))
-                    throw LogHelper.LogExceptionMessage(new OpenIdConnectProtocolException(LogHelper.FormatInvariant(LogMessages.IDX21314, LogHelper.MarkAsNonPII(JwtRegisteredClaimNames.Sub.ToLowerInvariant()), idToken)));
+                    throw LogHelper.LogExceptionMessage(new OpenIdConnectProtocolException(LogHelper.FormatInvariant(LogMessages.IDX21314, LogHelper.MarkAsNonPII(JwtRegisteredClaimNames.Sub.ToLowerInvariant()), idToken.ToString())));
```
A couple points about the ToString() usage in addtion to what I've said above:
1. Ideally we have one way people mark tokens at log time. Not great devEx to have one way if it's a string, one way if it's a `JwtSecurityToken`, etc. Easier to write a method once to take that cognative burden from the dev making the change.
2. Do you know why we have to rebuild this string? Shouldn't we store the base64 string for the token (at least in some cases) when it's passed in the constructor? @brentschmaltz do you know why this is the behavior?
3. Somewhat related to (2) above ideally we'd be disarming the string in the same way as for other parts of the code. That we have multiple places rebuilding tokens from their parts in code is a signal we may have more code duplication than we need.

TimHannMSFT reviewed (changes requested):
Thanks for the changes so far Xiao.

You mentioned to me you also had two questions for Brent/Team? Can you please add those as comments in the PR and we can resolve there?

victorm-hernandez commented:
@ciaozhang minor observation, the title of the PR is very descriptive but the commit message is not. Let's try to keep our commit messages readable and with enough context for anyone to understand a change. 

---
In reply to: [1526562219](https://github.com/AzureAD/azure-activedirectory-identitymodel-extensions-for-dotnet/pull/2064#issuecomment-1526562219) [](http://example.com/codeflow?ancestors=1526562219)

---
In reply to: [1526562219](https://github.com/AzureAD/azure-activedirectory-identitymodel-extensions-for-dotnet/pull/2064#issuecomment-1526562219) [](http://example.com/codeflow?ancestors=1526562219)

---
In reply to: [1526562219](https://github.com/AzureAD/azure-activedirectory-identitymodel-extensions-for-dotnet/pull/2064#issuecomment-1526562219) [](http://example.com/codeflow?ancestors=1526562219)

victorm-hernandez commented on src/Microsoft.IdentityModel.JsonWebTokens/JwtTokenUtilities.cs:
```diff
@@ -431,6 +431,17 @@ private static long ParseTimeValue(JToken jToken, string claimName)
 
+        internal static string PrepareSecurityArtifact(string securityArtifact)
+        {
+            if (IdentityModelEventSource.ShowPII)
```
>ShowPII [](http://example.com/codeflow?start=41&length=7)

####             if (IdentityModelEventSource.ShowPII)

---

[BUG] Isn't this the other way around? ShowPII = true means we dont modify the security artifact, otherwise we remove the signature. 

If you agree we should negate this if statement. #Resolved

TimHannMSFT replied:
I think part of the reason for confusion here is there's a lot of conflation of logic going on. This is set to only execute if PII is enabled because that's the only time the string will be logged (otherwise it's scrubbed). This can lead to difficult to debug problems down the line though. Also why I was favoring/suggesting following the pattern of using the decoration pattern (e.g. MarkAsNonPII) so the logic to scrub log lines is consolidated and not spread out which will make it easier to understand.

victorm-hernandez replied:
I completely agree on the decorator pattern, not using it will most likely require additional work to adapt to the other helper methods. 

If I understood your explanation we only cleanup the string when show PII is enabled so we we never leak the signature in the logs? If that is the case we shouldnt even have two execution flows and always strip the value out.

This method belongs to the LogHelper as well.

victorm-hernandez commented on test/Microsoft.IdentityModel.JsonWebTokens.Tests/JwtTokenUtilitiesTests.cs:
```diff
@@ -13,6 +16,52 @@ namespace Microsoft.IdentityModel.JsonWebTokens.Tests
+        private const string TestMessageOneParam = "This is the parameter: '{0}'.";
+
+        [Fact]
+        public void LogExceptionsWithStringJwe_JsonWebTokenHandler()
```
>LogExceptionsWithStringJwe_JsonWebTokenHandler [](http://example.com/codeflow?start=20&length=46)

####         public void LogExceptionsWithStringJwe_JsonWebTokenHandler()

---

You don't validate the negative case, when ShowPII is set to false. This could be done on the same test. #Closed

victorm-hernandez commented on src/Microsoft.IdentityModel.JsonWebTokens/JwtTokenUtilities.cs:
```diff
@@ -431,6 +431,17 @@ private static long ParseTimeValue(JToken jToken, string claimName)
             throw LogHelper.LogExceptionMessage(new FormatException(LogHelper.FormatInvariant(LogMessages.IDX14300, LogHelper.MarkAsNonPII(claimName), jToken.ToString(), LogHelper.MarkAsNonPII(typeof(long)))));
         }
 
+        internal static string PrepareSecurityArtifact(string securityArtifact)
```
>PrepareSecurityArtifact [](http://example.com/codeflow?start=31&length=23)

####         internal static string PrepareSecurityArtifact(string securityArtifact)

---

nit: In my opinion this method doesn't describe very well what its done, may I suggest something on the line of SanitizeForLog or RemovePII #Resolved

victorm-hernandez replied:
In addition it seems this method is better suited for the Microsoft.IdentityModel.Logging.LogHelper class.

TimHannMSFT commented on src/Microsoft.IdentityModel.Logging/SecurityArtifact.cs:
```diff
@@ -0,0 +1,35 @@
+        /// <summary>
+        /// Returns a string that represents the <see cref="Argument"/>.
+        /// </summary>
+        /// <returns><c>Null</c> if the <see cref="Argument"/> is <c>null</c>, otherwise calls <see cref="System.ValueType.ToString()"/> method of the <see cref="Argument"/>.</returns>
```
####         /// <returns><c>Null</c> if the <see cref="Argument"/> is <c>null</c>, otherwise calls <see cref="System.ValueType.ToString()"/> method of the <see cref="Argument"/>.</returns>

---

supernit: we use
```
<c>null</c>
```
 and
```
<see langword="null" />
```

We should standardize. I prefer the later because I think it produces better help syntax and I believe it's what .NET libraries do as well.
 #Resolved

victorm-hernandez reviewed (dismissed):
:clock1:

ciaozhang committed:
Introduce ISafeLogSecurityArtifact

rename the interface

temp save

temp save

santize logic

resolve comments

address comments

address comments

jmprieur commented on src/Microsoft.IdentityModel.JsonWebTokens/JsonWebTokenHandler.cs:
```diff
@@ -1238,13 +1240,13 @@ private static TokenValidationResult ReadToken(string token, TokenValidationPara
             if (validationParameters.TokenReader != null)
             {
-                var securityToken = validationParameters.TokenReader(token, validationParameters);
+                var securityToken = validationParameters.TokenReader(token, validationParameters); 
```
####                 var securityToken = validationParameters.TokenReader(token, validationParameters); 

---

extra space #Closed

ciaozhang replied:
Removed

jmprieur commented on src/Microsoft.IdentityModel.JsonWebTokens/JsonWebToken.cs:
```diff
@@ -579,14 +582,41 @@ private void ReadToken(string encodedJson)
+                sb.Append(EncryptedKey);
+                sb.Append('.');
+                sb.Append(InitializationVector);
+                sb.Append('.');
```
####                 sb.Append('.');

---

There are 4 dots in a JWT? I thought there were only 3? #Closed

ciaozhang replied:
There are 5 segments for JWE and 3 segments for a JWS.

jmprieur commented on src/Microsoft.IdentityModel.JsonWebTokens/JwtTokenUtilities.cs:
```diff
@@ -431,6 +431,17 @@ private static long ParseTimeValue(JToken jToken, string claimName)
+            if (IdentityModelEventSource.ShowPII)
+            {
+                if (!string.IsNullOrEmpty(securityArtifact) && (RegexJws.IsMatch(securityArtifact) || RegexJwe.IsMatch(securityArtifact)))
+                    return securityArtifact.Substring(0, securityArtifact.LastIndexOf("."));
```
####                     return securityArtifact.Substring(0, securityArtifact.LastIndexOf("."));

---

ah! back to the question 3 or 4? #Resolved

jmprieur commented on src/Microsoft.IdentityModel.Logging/IdentityModelEventSource.cs:
```diff
@@ -35,6 +35,11 @@ private IdentityModelEventSource()
+        /// <summary>
+        /// Flag which indicates whether or not armed security artifact is shown in logs when <see cref="ShowPII"/> is set to true. False by default.
+        /// </summary>
+        public static bool ShowArmedSecurityArtifact { get; set; } = false;
```
####         public static bool ShowArmedSecurityArtifact { get; set; } = false;

---

What is the end to end for libraries using it?
How does it integrate with the IIdentityLogger? #Resolved

jmprieur commented on src/Microsoft.IdentityModel.Logging/NonPII.cs:
```diff
@@ -26,10 +26,10 @@ public NonPII(object argument)
         public override string ToString()
         {
-            return Argument?.ToString() ?? "Null";
+            return Argument?.ToString() ?? "null";
```
####             return Argument?.ToString() ?? "null";

---

Isn't it going to mess-up existing logs? #Resolved

brentschmaltz replied:
+1 no need to make this change from Null => null.

ciaozhang replied:
Changed to "Null"

jmprieur reviewed (commented):
Thanks @ciaozhang 
I've asked questions.

Not sure why this is so complicated?
why not just:

if (PII)
{
 write token;
}
else
{
 write token.SubString(0, token.LastIndexOf('.'));
}

victorm-hernandez commented on src/Microsoft.IdentityModel.JsonWebTokens/JsonWebTokenHandler.cs:
```diff
@@ -3,7 +3,9 @@
 
 using System;
 using System.Collections.Generic;
+using System.IO;
```
>IO [](http://example.com/codeflow?start=13&length=2)

#### using System.IO;

---

Do you really need these 2 namespaces? From the code it seems there is no need. Could you confirm? #Closed

victorm-hernandez commented on src/Microsoft.IdentityModel.Logging/LogHelper.cs:
```diff
@@ -358,9 +360,42 @@ public static string FormatInvariant(string format, params object[] args)
+        /// <summary>
+        /// JWS - Token format: 'header.payload.signature'. Signature is optional, but '.' is required.
+        /// </summary>
+        public const string JsonCompactSerializationRegex = @"^[A-Za-z0-9-_]+\.[A-Za-z0-9-_]+\.[A-Za-z0-9-_]*$";
```
####         public const string JsonCompactSerializationRegex = @"^[A-Za-z0-9-_]+\.[A-Za-z0-9-_]+\.[A-Za-z0-9-_]*$";

---

All of this code is never referenced and is outside of concern of this namespace, lets remove the regex code. #Closed

victorm-hernandez commented on src/Microsoft.IdentityModel.Logging/LogHelper.cs:
```diff
@@ -358,9 +360,42 @@ public static string FormatInvariant(string format, params object[] args)
+
+        private static string SanitizeSecurityArtifact(object arg)
+        {
+            if (arg is ISafeLogSecurityArtifact)
```
>ISafeLogSecurityArtifact [](http://example.com/codeflow?start=23&length=24)

####             if (arg is ISafeLogSecurityArtifact)

---

If you are planning to use the ISafeLogSecurityArtifact interface, there is no need for the SecurityArtifact type and the related helper methods to wrap objects with this type. 
 #Resolved

henrik-me replied:
method could have the argument be of type SecurityArtifact instead. right now you are typecasting 3 places. and every path has 2 type casts

brentschmaltz replied:
Several places in our call graphs log tokens as strings, since they are not ISafeLogSecurityArtifact, ToString() will be called and the entire token with signature will be logged.

victorm-hernandez commented on src/Microsoft.IdentityModel.JsonWebTokens/JsonWebTokenHandler.cs:
```diff
@@ -1238,13 +1240,13 @@ private static TokenValidationResult ReadToken(string token, TokenValidationPara
+                var securityToken = validationParameters.TokenReader(token, validationParameters); 
                 if (securityToken == null)
-                    throw LogHelper.LogExceptionMessage(new SecurityTokenInvalidSignatureException(LogHelper.FormatInvariant(TokenLogMessages.IDX10510, token)));
+                    throw LogHelper.LogExceptionMessage(new SecurityTokenInvalidSignatureException(LogHelper.FormatInvariant(TokenLogMessages.IDX10510, LogHelper.MarkAsSecurityArtifact(token))));
```
>MarkAsSecurityArtifact [](http://example.com/codeflow?start=162&length=22)

####                     throw LogHelper.LogExceptionMessage(new SecurityTokenInvalidSignatureException(LogHelper.FormatInvariant(TokenLogMessages.IDX10510, LogHelper.MarkAsSecurityArtifact(token))));

---

This method wraps the argument in a SecurityArtifact object, yet this type is not used anywhere on the LogHelper. Since the current implementation relies on the ISafeLogSecurityArtifact interface there is no need for this helper method and the SecurityArtifact type. 

 #Resolved

ciaozhang replied:
We only implement the interface for the object derived from SecurityToken. Here we log the token as string that the LogHelper cannot scrub the signature for a string.

victorm-hernandez reviewed (dismissed):
:clock1:

henrik-me commented on src/Microsoft.IdentityModel.Logging/ISafeLogSecurityArtifact.cs:
```diff
@@ -0,0 +1,22 @@
+    public interface ISafeLogSecurityArtifact
+    {
+        /// <summary>
+        /// Returns a string that represents the disarmed security artifact.
```
>disarmed [](http://example.com/codeflow?start=49&length=8)

####         /// Returns a string that represents the disarmed security artifact.

---

what is a disarmed security artifact? and what is an armed security artifact? ... and also we are explaining the method names with words that are even harder to understand. Can we pls. simplify the message and also, can we simplify the names of these 2 methods - it's not logical to me what these 2 things are doing? #Resolved

henrik-me commented on src/System.IdentityModel.Tokens.Jwt/JwtSecurityToken.cs:
```diff
@@ -469,6 +469,18 @@ public override string ToString()
         }
 
+        /// <inheritdoc/>
+        public override string SecureToString()
```
>SecureToString [](http://example.com/codeflow?start=31&length=14)

####         public override string SecureToString()

---

Why not simply call these what they do? 
NoSignatureToString() and WithSignatureToString() #Resolved

brentschmaltz replied:
For that matter perhaps a property: LogHelper.LogTokensWithSignature.

brentschmaltz commented on src/Microsoft.IdentityModel.Logging/LogHelper.cs:
```diff
@@ -358,9 +360,42 @@ public static string FormatInvariant(string format, params object[] args)
+        /// <summary>
+        /// Regex that is used to figure out if a token is in JWS format.
+        /// </summary>
+        public static readonly Regex RegexJws = new Regex(JsonCompactSerializationRegex, RegexOptions.Compiled | RegexOptions.CultureInvariant, TimeSpan.FromMilliseconds(100));
```
####         public static readonly Regex RegexJws = new Regex(JsonCompactSerializationRegex, RegexOptions.Compiled | RegexOptions.CultureInvariant, TimeSpan.FromMilliseconds(100));

---

These should not be public. #Resolved

brentschmaltz commented on src/Microsoft.IdentityModel.JsonWebTokens/JwtTokenUtilities.cs:
```diff
@@ -431,6 +431,12 @@ private static long ParseTimeValue(JToken jToken, string claimName)
 
+        internal static string SafeLogJwtToken(object obj)
+        {
+            string token = obj as string;
```
####             string token = obj as string;

---

Suggested logic here

```cs
if (obj == null)
    return string.Empty;

// not a string
if (token == null)
    return obj.ToString()

int lastDot = token.LastIndexOf(".");

// no dots, maybe not a JWT
if (lastDot == -1)
    return typeof(string).ToString();

return token.Substring(0,lastDot);

``` #Resolved

ciaozhang replied:
Done

victorm-hernandez commented on src/Microsoft.IdentityModel.Logging/ISafeLogSecurityArtifact.cs:
```diff
@@ -0,0 +1,18 @@
+    {
+        /// <summary>
+        /// Returns a string that represents the complete security artifact.
+        /// This may include sensitive information and should only be used for logging purposes.
```
>logging [](http://example.com/codeflow?start=79&length=7)

We should not encourage the use of this for logging.

This should be used for exceptional debugging cases as it will reveal security tokens in plain text. 

 #Resolved

victorm-hernandez replied:
@ciaozhang

victorm-hernandez replied:
Still waiting for this change @ciaozhang

victorm-hernandez commented on src/Microsoft.IdentityModel.JsonWebTokens/JsonWebToken.cs:
```diff
@@ -891,6 +894,15 @@ public string Subject
+        /// <summary>
+        /// Returns the encoded token without signature or authentication tag.
+        /// </summary>
+        /// <returns>Encoded token string  without signature or authentication tag.</returns>
```
>   [](http://example.com/codeflow?start=41&length=2)

####         /// <returns>Encoded token string  without signature or authentication tag.</returns>

---

nit: extra space #Closed

ciaozhang replied:
Fixed

victorm-hernandez commented on src/Microsoft.IdentityModel.Logging/SecurityArtifact.cs:
```diff
@@ -0,0 +1,53 @@
+{
+    /// <summary>
+    /// An internal structure that is used to mark an argument as SecurityArtifact.
+    /// Arguments wrapped with a SecurityArtifact structure will be considered as SecurityArtifact in the message logging process.
```
>Arguments [](http://example.com/codeflow?start=8&length=9)

####     /// Arguments wrapped with a SecurityArtifact structure will be considered as SecurityArtifact in the message logging process.

---

This may be the right place to clarify that this is meant for the scenario where the argument is not a class but a simple string. If the argument is a string there is no way to add an interface that can be used by the LogHelper.

Instead we wrap the string with this class and pass a delegate that knows how to sanitize or not the string. #Closed

victorm-hernandez commented on src/Microsoft.IdentityModel.Logging/SecurityArtifact.cs:
```diff
@@ -0,0 +1,53 @@
+            return Argument?.ToString() ?? "null";
+        }
+
+        public string UnsafeToString()
```
>UnsafeToString [](http://example.com/codeflow?start=22&length=14)

XML docs missing. #Resolved

ciaozhang replied:
Added

victorm-hernandez commented on src/Microsoft.IdentityModel.Protocols.OpenIdConnect/OpenIdConnectProtocolValidator.cs:
```diff
@@ -511,13 +511,13 @@ protected virtual void ValidateCHash(OpenIdConnectProtocolValidationContext vali
             if (!validationContext.ValidatedIdToken.Payload.TryGetValue(JwtRegisteredClaimNames.CHash, out cHashClaim))
             {
-                throw LogHelper.LogExceptionMessage(new OpenIdConnectProtocolInvalidCHashException(LogHelper.FormatInvariant(LogMessages.IDX21307, validationContext.ValidatedIdToken)));
+                throw LogHelper.LogExceptionMessage(new OpenIdConnectProtocolInvalidCHashException(LogHelper.FormatInvariant(LogMessages.IDX21307, validationContext.ValidatedIdToken.ToString())));
```
>ToString [](http://example.com/codeflow?start=182&length=8)

Why do you convert to a string here? Isn't this the opposite of what we want? And if you make it a string, should not you use the MarkAsSecurityArtifact?

ciaozhang replied:
Removed ToString()

brentschmaltz replied:
+1

victorm-hernandez commented on src/System.IdentityModel.Tokens.Jwt/JwtSecurityToken.cs:
```diff
@@ -469,6 +469,9 @@ public override string ToString()
         }
 
+        /// <inheritdoc/>
+        public override string UnsafeToString() => RawData;
```
>UnsafeToString [](http://example.com/codeflow?start=31&length=14)

Lets update the Saml2SecurityToken as well to have parity on this feature. #Resolved

brentschmaltz replied:
@victorm-hernandez are you with tackling SAML after the JWT's.

brentschmaltz replied:
@victorm-hernandez can we open a new item for Saml as removing the signature from Saml tokens is a bit more complicated that JWT's and will need some thought.

victorm-hernandez replied:
Agreed @brentschmaltz , lets do this in another PR. @ciaozhang could you please create the work item and leave it here.

ciaozhang replied:
WI created:
https://github.com/AzureAD/azure-activedirectory-identitymodel-extensions-for-dotnet/issues/2090

victorm-hernandez replied:
Created ADO work item as well
Product Backlog Item 2595096: Replicate the logic to "disarm security artifact" from JWT tokens to SAML tokens

victorm-hernandez reviewed (changes requested):
:clock1:

ciaozhang committed:
address comment

ciaozhang committed:
address comment

ciaozhang committed:
fix failed test

victorm-hernandez commented on src/Microsoft.IdentityModel.JsonWebTokens/JwtTokenUtilities.cs:
```diff
@@ -429,6 +429,24 @@ private static long ParseTimeValue(JToken jToken, string claimName)
             throw LogHelper.LogExceptionMessage(new FormatException(LogHelper.FormatInvariant(LogMessages.IDX14300, LogHelper.MarkAsNonPII(claimName), jToken.ToString(), LogHelper.MarkAsNonPII(typeof(long)))));
         }
 
+        internal static string SafeLogJwtToken(object obj)
```
>object [](http://example.com/codeflow?start=47&length=6)

I thought this was meant for strings only, for any other object type we should be able to use the provided interface #Closed

victorm-hernandez replied:
@ciaozhang

ciaozhang replied:
It is the callback func we used in SecurityArtifact struct. The reason we take obj as the type of parameter is LogHelper treat all arguments as object.

victorm-hernandez replied:
I see, still regardless of the parameter type this is meant to be consumed by using string version of tokens correct? 

how do you justify the code for non strings that you have here?

ciaozhang replied:
It will use the default ToString() method return the GetType().ToString()

victorm-hernandez commented on src/Microsoft.IdentityModel.JsonWebTokens/JwtTokenUtilities.cs:
```diff
@@ -429,6 +429,24 @@ private static long ParseTimeValue(JToken jToken, string claimName)
+
+            // not a string
+            if (!(obj is string token))
+                return obj.ToString();
```
This doesn't seem safe at all. By doing this we are giving the impression of safety while we are not really doing any changes. #Closed

ciaozhang replied:
We only mark string as Security artifact.

victorm-hernandez replied:
What I mean here is that if I call your method I should be able to assume that your output is now safe. In the case of non strings you cannot guarantee that, the only safe choice is to return null here or string.empty.

victorm-hernandez replied:
Since this is really an exceptional case and should never happen you may consider throwing.

ciaozhang replied:
return "UncognizedEncodedString"

victorm-hernandez commented on src/Microsoft.IdentityModel.JsonWebTokens/JwtTokenUtilities.cs:
```diff
@@ -429,6 +429,24 @@ private static long ParseTimeValue(JToken jToken, string claimName)
+            int lastDot = token.LastIndexOf(".");
+
+            // no dots, maybe not a JWT
+            if (lastDot == -1)
```
>lastDot [](http://example.com/codeflow?start=16&length=7)

Since this is meant to parse JWT token strings we should throw an argument exception here.  #Closed

ciaozhang replied:
We will throw in the ReadToken() method if the string token is not an encoded JWT token.

victorm-hernandez replied:
This should be safe regardless of the caller, for non strings, non tokens we should return null.

victorm-hernandez commented on src/Microsoft.IdentityModel.Logging/IdentityModelEventSource.cs:
```diff
@@ -35,6 +35,11 @@ private IdentityModelEventSource()
         public static bool ShowPII { get; set; } = false;
 
+        /// <summary>
+        /// Flag which indicates whether or not armed security artifact is shown in logs when <see cref="ShowPII"/> is set to true. False by default.
```
>true [](http://example.com/codeflow?start=126&length=4)

It is not clear for someone without context what a security artifact is. Let's provide a concrete example, security tokens  #Closed

victorm-hernandez replied:
@ciaozhang

ciaozhang replied:
Updated the XML.

ciaozhang committed:
update xml

victorm-hernandez commented on src/Microsoft.IdentityModel.Logging/ISafeLogSecurityArtifact.cs:
```diff
@@ -0,0 +1,18 @@
+namespace Microsoft.IdentityModel.Logging
+{
+    /// <summary>
+    /// Interface that provides an unsafe method to log a security artifact.
```
>Interface [](http://example.com/codeflow?start=8&length=9)

You may want to bring the description you have for the SecurityArtifact class to the interface, this would address the questions that Henrik had here.  #Resolved

victorm-hernandez commented on src/Microsoft.IdentityModel.Logging/LogHelper.cs:
```diff
@@ -393,6 +406,17 @@ public static object MarkAsNonPII(object arg)
+        /// <param name="arg">A log message argument to be marked as SecurityArtifact.</param>
+        /// <param name="callback">A callback function to log the security artifact safely.</param>
+        /// <returns>An argument marked as SecurityArtifact.</returns>
+        public static object MarkAsSecurityArtifact(object arg, Func<object, string> callback)
```
Why do we receive an object here, this should receive and return a string.  #Resolved

ciaozhang committed:
address comments

ciaozhang committed:
address comments

ciaozhang committed:
address comments

victorm-hernandez reviewed (approved):
:shipit:

TimHannMSFT reviewed (approved):
Approving based on Victor's approval and seeing that the new direction is more consistent. Thanks for the changes @ciaozhang!
